File: 5.问题五贪心+遗传/core_objects.cpp
#include "core_objects.hpp"
#include <cmath>
#include <stdexcept>
#include <algorithm>
#include <vector>

namespace CoreObjects {

// TargetCylinder Implementation
TargetCylinder::TargetCylinder(const Config::TargetSpecs& specs, 
                               int num_circ_samples, 
                               int num_height_samples)
    : radius_(specs.radius)
    , height_(specs.height)
    , bottom_center_(specs.center_bottom)
    , top_center_(specs.center_bottom + Vector3d(0.0, 0.0, specs.height))
{
    generate_full_key_points(num_circ_samples, num_height_samples);
}

void TargetCylinder::generate_full_key_points(int num_circ_samples, int num_height_samples) {
    std::vector<Vector3d> points;
    
    // 1. 上下底面圆盘的采样 (圆周 + 圆心)
    points.push_back(bottom_center_);
    points.push_back(top_center_);
    
    for (int i = 0; i < num_circ_samples; ++i) {
        double angle = 2.0 * M_PI * i / num_circ_samples;
        Vector3d offset_xy(radius_ * std::cos(angle), radius_ * std::sin(angle), 0.0);
        
        // 底面圆周点
        points.push_back(bottom_center_ + offset_xy);
        // 顶面圆周点
        points.push_back(top_center_ + offset_xy);
    }
    
    // 2. 侧面母线的采样
    constexpr int num_side_samples = 4;
    for (int i = 0; i < num_side_samples; ++i) {
        double angle = 2.0 * M_PI * i / num_side_samples;
        Vector3d offset_xy(radius_ * std::cos(angle), radius_ * std::sin(angle), 0.0);
        
        // 在母线上从下到上均匀取点 (不含端点，因为已被圆周覆盖)
        for (int j = 1; j < num_height_samples; ++j) {
            double height_fraction = static_cast<double>(j) / num_height_samples;
            Vector3d height_offset(0.0, 0.0, height_ * height_fraction);
            points.push_back(bottom_center_ + offset_xy + height_offset);
        }
    }
    
    // 转换为Eigen矩阵格式 (3xN)
    key_points_.resize(3, points.size());
    for (size_t i = 0; i < points.size(); ++i) {
        key_points_.col(i) = points[i];
    }
}

// Missile Implementation
Missile::Missile(const std::string& missile_id) : id_(missile_id) {
    auto it = Config::MISSILES_INITIAL.find(missile_id);
    if (it == Config::MISSILES_INITIAL.end()) {
        throw std::runtime_error("Unknown missile ID: " + missile_id);
    }
    
    const auto& specs = it->second;
    start_pos_ = specs.pos;
    speed_ = specs.speed;
    
    Vector3d direction_vec = specs.target - start_pos_;
    unit_vec_ = direction_vec.normalized();
}

Vector3d Missile::get_position(double t) const {
    return start_pos_ + unit_vec_ * speed_ * t;
}

// SmokeCloud Implementation
SmokeCloud::SmokeCloud(const Vector3d& detonate_pos, double detonate_time)
    : detonate_pos_(detonate_pos)
    , start_time_(detonate_time)
    , end_time_(detonate_time + Config::CLOUD_DURATION)
{
}

std::optional<Vector3d> SmokeCloud::get_center(double t) const {
    if (t < start_time_ || t >= end_time_) {
        return std::nullopt;
    }
    
    double t_since_detonate = t - start_time_;
    Vector3d sink_offset(0.0, 0.0, -Config::CLOUD_SINK_SPEED * t_since_detonate);
    return detonate_pos_ + sink_offset;
}

// TrajectoryIntegrator Implementation
Vector3d TrajectoryIntegrator::solve_trajectory(
    const Vector3d& deploy_pos,
    const Vector3d& deploy_vel, 
    double fuse_time,
    double mass,
    double drag_factor
) {
    // 使用4阶Runge-Kutta方法求解ODE
    // 状态向量: [x, y, z, vx, vy, vz]
    Eigen::VectorXd y(6);
    y << deploy_pos[0], deploy_pos[1], deploy_pos[2], 
         deploy_vel[0], deploy_vel[1], deploy_vel[2];
    
    double t = 0.0;
    double dt = 0.01; // 时间步长
    
    while (t < fuse_time) {
        double h = std::min(dt, fuse_time - t);
        
        // RK4积分步骤
        Eigen::VectorXd k1(6), k2(6), k3(6), k4(6);
        Eigen::VectorXd y_temp(6);
        
        grenade_motion_ode(t, y, k1, mass, drag_factor);
        
        y_temp = y + 0.5 * h * k1;
        grenade_motion_ode(t + 0.5*h, y_temp, k2, mass, drag_factor);
        
        y_temp = y + 0.5 * h * k2;
        grenade_motion_ode(t + 0.5*h, y_temp, k3, mass, drag_factor);
        
        y_temp = y + h * k3;
        grenade_motion_ode(t + h, y_temp, k4, mass, drag_factor);
        
        y += h/6.0 * (k1 + 2*k2 + 2*k3 + k4);
        t += h;
    }
    
    return Vector3d(y[0], y[1], y[2]);
}

void TrajectoryIntegrator::grenade_motion_ode(
    double t,
    const Eigen::VectorXd& y,
    Eigen::VectorXd& dydt,
    double mass,
    double drag_factor
) {
    // y[0:3] 是位置, y[3:6] 是速度
    Vector3d velocity(y[3], y[4], y[5]);
    
    // 计算加速度
    Vector3d gravity_accel(0.0, 0.0, -Config::G);
    
    double speed = velocity.norm();
    Vector3d drag_accel = Vector3d::Zero();
    if (speed > 1e-6) {
        drag_accel = -(drag_factor / mass) * speed * velocity;
    }
    
    Vector3d total_accel = gravity_accel + drag_accel;
    
    // 返回状态向量的导数 [d(pos)/dt, d(vel)/dt] = [vel, accel]
    dydt << velocity[0], velocity[1], velocity[2],
            total_accel[0], total_accel[1], total_accel[2];
}

// Grenade Implementation
Grenade::Grenade(const Vector3d& deploy_pos, 
                 const Vector3d& deploy_vel,
                 double deploy_time, 
                 double fuse_time)
    : deploy_time_(deploy_time)
    , fuse_time_(fuse_time)
    , detonate_time_(deploy_time + fuse_time)
{
    detonate_pos_ = TrajectoryIntegrator::solve_trajectory(deploy_pos, deploy_vel, fuse_time);
}

std::unique_ptr<SmokeCloud> Grenade::generate_smoke_cloud() const {
    return std::make_unique<SmokeCloud>(detonate_pos_, detonate_time_);
}

// UAV Implementation
UAV::UAV(const std::string& uav_id) 
    : id_(uav_id), strategy_set_(false) 
{
    auto it = Config::UAVS_INITIAL.find(uav_id);
    if (it == Config::UAVS_INITIAL.end()) {
        throw std::runtime_error("Unknown UAV ID: " + uav_id);
    }
    
    start_pos_ = it->second.pos;
}

void UAV::set_flight_strategy(double speed, double angle) {
    speed_ = speed;
    angle_ = angle;
    velocity_vec_ = speed * Vector3d(std::cos(angle), std::sin(angle), 0.0);
    strategy_set_ = true;
}

Vector3d UAV::get_position(double t) const {
    if (!strategy_set_) {
        throw std::runtime_error("UAV flight strategy has not been set.");
    }
    return start_pos_ + velocity_vec_ * t;
}

std::unique_ptr<Grenade> UAV::deploy_grenade(double deploy_time, double fuse_time) const {
    if (!strategy_set_) {
        throw std::runtime_error("UAV flight strategy has not been set.");
    }
    
    Vector3d deploy_pos = get_position(deploy_time);
    return std::make_unique<Grenade>(deploy_pos, velocity_vec_, deploy_time, fuse_time);
}

} // namespace CoreObjects

File: 3.解答代码_solve1_2/core_objects.py
# core_objects.py
import numpy as np
from config import *
from geometry import check_full_obscuration
from scipy.integrate import solve_ivp
class TargetCylinder:
    def __init__(self, specs):
        self.radius = specs['radius']
        self.height = specs['height']
        self.bottom_center = specs['center_bottom']
        self.top_center = self.bottom_center + np.array([0.0, 0.0, self.height])

class Missile:
    def __init__(self, missile_id):
        self.id = missile_id
        specs = MISSILES_INITIAL[missile_id]
        self.start_pos = specs['pos']
        self.speed = specs['speed']
        direction_vec = specs['target'] - self.start_pos
        self.unit_vec = direction_vec / np.linalg.norm(direction_vec)

    def get_position(self, t):
        return self.start_pos + self.unit_vec * self.speed * t

class SmokeCloud:
    def __init__(self, detonate_pos, detonate_time):
        self.detonate_pos = detonate_pos
        self.start_time = detonate_time
        self.end_time = detonate_time + CLOUD_DURATION

    def get_center(self, t):
        if not (self.start_time <= t < self.end_time):
            return None
        t_since_detonate = t - self.start_time
        return self.detonate_pos - np.array([0.0, 0.0, CLOUD_SINK_SPEED * t_since_detonate])
    
    def check_obscuration_at_time(self, t, missile, target):
        missile_pos = missile.get_position(t)
        cloud_center = self.get_center(t)
        if cloud_center is None:
            return False
        return check_full_obscuration(missile_pos, cloud_center, target)

def grenade_motion_ode(t, y, mass, drag_factor):
    """
    定义烟幕弹运动的常微分方程组。
    y 是一个6维状态向量: [x, y, z, vx, vy, vz]
    """
    # y[0:3] 是位置, y[3:6] 是速度
    velocity = y[3:6]
    
    # 计算加速度
    gravity_accel = np.array([0.0, 0.0, -G])
    
    speed = np.linalg.norm(velocity)
    if speed > 1e-6:
        drag_accel = -(drag_factor / mass) * speed * velocity
    else:
        drag_accel = np.array([0.0, 0.0, 0.0])
        
    total_accel = gravity_accel + drag_accel
    
    # 返回状态向量的导数 [d(pos)/dt, d(vel)/dt] = [vel, accel]
    return np.concatenate((velocity, total_accel))

class Grenade:
    def __init__(self, deploy_pos, deploy_vel, deploy_time, fuse_time):
        self.deploy_time = deploy_time
        self.fuse_time = fuse_time
        self.detonate_time = self.deploy_time + self.fuse_time
        
        # 使用专业的ODE求解器计算起爆点
        self.detonate_pos = self._solve_trajectory_scipy(deploy_pos, deploy_vel)

    def _solve_trajectory_scipy(self, deploy_pos, deploy_vel):
        """
        使用 scipy.integrate.solve_ivp 求解弹道。
        """
        # 初始状态向量 [x, y, z, vx, vy, vz]
        y0 = np.concatenate((deploy_pos, deploy_vel))
        
        # 求解时间范围 [0, fuse_time]
        t_span = [0, self.fuse_time]
        
        # 我们只关心最终时刻的结果，所以设置 t_eval=[self.fuse_time]
        solution = solve_ivp(
            fun=grenade_motion_ode,      # 微分方程
            t_span=t_span,               # 时间范围
            y0=y0,                       # 初始条件
            method='RK45',               # 求解器方法 (常用且高效)
            t_eval=[self.fuse_time],     # 指定输出时间的点
            args=(GRENADE_MASS, GRENADE_DRAG_FACTOR) # 传递给 fun 的额外参数
        )
        
        # 提取最终时刻的位置
        final_state = solution.y[:, -1]
        final_position = final_state[0:3]
        
        return final_position
    
    def generate_smoke_cloud(self):
        return SmokeCloud(self.detonate_pos, self.detonate_time)
    
class UAV:
    def __init__(self, uav_id):
        self.id = uav_id
        self.start_pos = UAVS_INITIAL[uav_id]['pos']
        self.speed = None
        self.angle = None
        self.velocity_vec = None

    def set_flight_strategy(self, speed, angle):
        self.speed = speed
        self.angle = angle
        self.velocity_vec = speed * np.array([np.cos(angle), np.sin(angle), 0.0])
    
    def get_position(self, t):
        if self.velocity_vec is None:
            raise ValueError("UAV flight strategy has not been set.")
        return self.start_pos + self.velocity_vec * t
    
    def deploy_grenade(self, deploy_time, fuse_time):
        if self.velocity_vec is None:
            raise ValueError("UAV flight strategy has not been set.")
        deploy_pos = self.get_position(deploy_time)
        return Grenade(deploy_pos, self.velocity_vec, deploy_time, fuse_time)
